src/main.c

```diff
@@ -166,8 +166,31 @@ int main(void) {
 #elif defined(SAMD51)
     WDT->CTRLA.reg = 0;
     while(WDT->SYNCBUSY.reg) {}
+
+    // Enable 2.7V brownout detection. The default fuse value is 1.7
+    // Set brownout detection to ~2.7V. Default from factory is 1.7V,
+    // which is too low for proper operation of external SPI flash chips (they are 2.7-3.6V).
+    // Also without this higher level, the SAMD51 will write zeros to flash intermittently.
+    // Disable while changing level.
+    SUPC->BOD33.bit.ENABLE = 0;
+    SUPC->BOD33.bit.LEVEL = 200;  // 2.7V: 1.5V + LEVEL * 6mV.
+    // Don't reset right now.
+    SUPC->BOD33.bit.ACTION = SUPC_BOD33_ACTION_NONE_Val;
+    SUPC->BOD33.bit.ENABLE = 1; // enable brown-out detection
+
+    while (SUPC->STATUS.bit.BOD33DET) {
+        // Wait for voltage to rise above BOD33 value.
```

Comment:
This will busy wait the CPU. Do you want to use the BOD33 to do the wake up? You could set the action to interrupt and then WFI. That way you won't burn as much power while a battery charges. (You could also use WAKEEN and the lower power modes too.)

Reply:
That's a good possibility. I have to configure the BOD33 to be active in a certain sleep mode. Based on your newfound knowledge of sleeping, do you have a suggestion of whether to set `SLEEPCFG.SLEEPMODE` to `IDLE` or to `STANDBY`?

I don't even need a real interrupt handler because I can just check the BOD33 status after the WFI completes, and repeat if it's still too low.

Reply:
Hmm, it turns out there is no interrupt that fires when the voltage has _risen_ to the specified level; the interrupt available is only when it falls below the level.

Reply:
An alternative would be to sleep for a short while and re-check, but that requires enabling the low-power clocks, etc. I'd like to defer doing that until later, so that we have a solid bootloader that fixes the fundamental problem. Later we can add better low-power handling. The current busy-wait is no worse than the repeated reset that would happen at the edge of the BOD33 value.

I'm running a multi-hour test right now; it's doing fine.

Reply:
Probably not worth the complexity at this point. Two options would be:

1) Use the RTC or a timer to sleep a fixed duration between checks.
2) Use an analog comparator to monitor the voltage and wake us up.

It doesn't look like we can use the SUPC to wake us up because the VBAT isn't connected and therefore we can't switch to and from it.